### PR TITLE
Research

### PR DESCRIPTION
**Pull request by AlexeyKozhevin**
- Speedify Research by changes in saving
- Fix usage of callables with the same name
- Fix artifacts load
- Fix research profiler
- Minor changes: `__str__`, column renaming in research profiler, ...
- Extend tests

### CODE REVIEW

File: batchflow/research/experiment.py [hunk 1]
```diff
@@ -53,11 +56,11 @@ def __call__(self, config, batch=None, **kwargs):
...
         Returns
         -------
-        generator, Pipeline or Batch
+        list or generator
             return depends on the mode:
                 - 'generator': generator
-                - 'func': Pipeline itself
-                - 'execute_for': processed batch
+                - 'func': Pipeline itself and variables values
+                - 'execute_for': processed batch and variables values
```
Comment (a-arefina) [inline batchflow/research/experiment.py:63]:
```suggestion
                - 'func':  the pipeline object and its variables values
                - 'execute_for': the processed batch and its variables values
```

File: batchflow/research/experiment.py [hunk 2]
```diff
@@ -53,11 +56,11 @@ def __call__(self, config, batch=None, **kwargs):
 
         Returns
         -------
-        generator, Pipeline or Batch
+        list or generator
```
Comment (a-arefina) [inline batchflow/research/experiment.py:59]:
```suggestion
        tuple or generator
```

File: batchflow/research/experiment.py [hunk 3]
```diff
@@ -566,15 +599,21 @@ def dump(self, variable=None, when='last'):
...
+        name = self.add_postfix(name)
         self.add_callable(name, _dump_results,
                           when=when,
                           variable=variable, experiment=E())
         return self
 
     def add_postfix(self, name):
         """ Add postfix for conincided unit name. """
-        return name + '_' + str(sum([item.startswith(name) for item in self.actions]))
+        n_actions = sum([self._is_postifixed(name, item) for item in self.actions])
```
Comment (a-arefina) [inline batchflow/research/experiment.py:611]: you can drop brackets and skip intermediate list creation

File: batchflow/research/profiler.py [hunk 4]
```diff
@@ -44,8 +47,8 @@ def show_profile_info(self, per_iter=False, per_experiment=False, detailed=False
         detailed = False if not self.detailed else detailed
 
         if per_iter is False and detailed is False:
-            groupby = ['experiment', 'action', 'iter'] if per_experiment else ['action', 'iter']
-            columns = columns or ['total_time', 'eval_time']
+            groupby = ['experiment', 'unit', 'iter'] if per_experiment else ['unit', 'iter']
```
Comment (a-arefina) [inline batchflow/research/profiler.py:50]: did you mean self.UNIT ?

File: batchflow/research/profiler.py [hunk 5]
```diff
@@ -56,33 +59,34 @@ def show_profile_info(self, per_iter=False, per_experiment=False, detailed=False
                 result = result.sort_index(level=0)
 
         elif per_iter is False and detailed is True:
-            groupby = ['experiment', 'action', 'id'] if per_experiment else ['action', 'id']
+            groupby = ['experiment', 'unit', 'id'] if per_experiment else ['unit', 'id']
```
Comment (a-arefina) [inline batchflow/research/profiler.py:62]: I would prefer ` groupby = (['experiment']  if per_experiment else []) + ['unit', 'id']` to omit repeating the second term, but this is really up to you

File: batchflow/tests/research_test.py [hunk 6]
```diff
@@ -327,15 +327,18 @@ def inc(x):
         executor = Executor(experiment, target='f', configs=[{'n': 10}, {'n': 20}], n_iters=None)
         executor.run()
 
-    @pytest.mark.parametrize('save_to', ['a', ['a', 'b', 'c']])
+    @pytest.mark.parametrize('save_to', ['a', ['a', 'b'], ['a', 'b', 'c']])
```
Comment (a-arefina) [inline batchflow/tests/research_test.py:330]: add ['a'] ?

**Review by a-arefina — APPROVED** [review]
Can't wait for seeing this PR in master!

See some minor comments

File: batchflow/research/profiler.py (hunk 4 above)
Comment (AlexeyKozhevin) [reply batchflow/research/profiler.py:50]: `self.UNIT_NAME` is for `_add_profile_info` which is common for all profilers.
We define `show_profile_info` for pipelines and experiments separately so we can write here `'unit'` directly.

**Pull request merged**